src/qibolab/instruments/icarusq.py

```diff
@@ -1,167 +1,30 @@
-        end = max(pulse.start + pulse.duration for pulse in sequence)
-        time_array = np.arange(
-            start * 1e-9 - self.pulse_buffer,
-            end * 1e-9 + self.pulse_buffer,
-            1 / self.sample_rate,
-        )
-        waveform_arrays = np.zeros((self.device.num_channels, len(time_array)))
-
-        for pulse in sequence:
-            start_index = bisect(time_array, pulse.start * 1e-9)
-            end_index = bisect(time_array, (pulse.start + pulse.duration) * 1e-9)
-            i_ch, q_ch = pulse.channel
-            i, q = self.generate_waveforms_from_pulse(pulse, time_array[start_index:end_index])
-            waveform_arrays[i_ch, start_index:end_index] += i
-            waveform_arrays[q_ch, start_index:end_index] += q
-
-        return waveform_arrays
-
-    def upload(self, waveform: np.ndarray):
-        """Uploads a nchannels X nsamples array to the AWG, load it into memory and assign it to the channels for playback."""
-
-        # TODO: Add additional check to ensure all waveforms are of the same size? Should be caught by qcodes driver anyway.
-        if len(waveform) != self.device.num_channels:
-            raise_error(Exception, "Invalid waveform given")
-
-        # Clear existing waveforms in memory
-        self.device.write("WLIS:WAV:DEL ALL")
-
-        # Upload waveform, load into memory and assign to each channel
-        for idx, channel in enumerate(range(1, self.device.num_channels + 1)):
-            awg_ch = getattr(self.device, f"ch{channel}")
-            wfmx = self.device.makeWFMXFile(waveform[idx], awg_ch.awg_amplitude())
-            self.device.sendWFMXFile(wfmx, f"ch{channel}.wfmx")
-            self.device.loadWFMXFile(f"ch{channel}.wfmx")
-            self.device.write(f'SOURce{channel}:CASSet:WAVeform "ch{channel}"')
-
-    def start(self):
         pass
 
-    def play_sequence(self):
-        for channel in range(1, self.device.num_channels + 1):
-            awg_ch = getattr(self.device, f"ch{channel}")
-            awg_ch.state(1)
-            self.device.write(f"SOURce{channel}:RMODe TRIGgered")
-            self.device.write(f"SOURce{channel}:TINPut ATR")
-        self.device.play()
-        self.device.wait_for_operation_to_complete()
-
-    def stop(self):
-        self.device.stop()
-        for channel in range(1, self.device.num_channels + 1):
-            awg_ch = getattr(self.device, f"ch{channel}")
-            awg_ch.state(0)
-        self.device.wait_for_operation_to_complete()
-
-    def disconnect(self):
-        if self.is_connected:
-            self.device.stop()
-            self.device.close()
-            self.is_connected = False
-
-    def __del__(self):
-        self.disconnect()
+    def play(self, *args):
+        pass
 
-    def close(self):
-        if self.is_connected:
-            self.stop()
-            self.device.close()
-            self.is_connected = False
+    @property
+    def attenuation(self):
+        import urllib3
```

Comment:
I will add it as a dependency, I need specifically this library for this instrument.

Reply:
I have tried with the other libraries three years ago but somehow only this one worked. 

```
>>> address = "192.168.0.9:90"
>>> res = requests.get(f"http://{address}/GETATT?")
>>> res.text
'-99 Unrecognized Command. MN:RCDAT-8000-30 SN:11611130010'
res = http.request("GET", f"http://{address}/GETATT?")
>>> res._body
b'30.0'
```

Reply:
I tried to check the difference between the two requests (spawning a simple server and handling them) and it seems it is only a question mark:

```py
# requests
(Pdb++) self.requestline
'GET /GETATT HTTP/1.1'
(Pdb++) self.headers.as_string()
'Host: 127.0.0.1:8080\nUser-Agent: python-requests/2.31.0\nAccept-Encoding: gzip, deflate\nAccept: */*\nConnection: keep-alive\n\n'
# urrlib3
(Pdb++) self.requestline
'GET /GETATT? HTTP/1.1'
(Pdb++) self.headers.as_string()
'Host: 127.0.0.1:8080\nAccept-Encoding: identity\nUser-Agent: python-urllib3/2.1.0\n\n'
```

I will check if it's possible to preserve the question mark in `requests` as well (even though it seems strange that makes such a difference for your HTTP server...)

Reply:
Never mind: despite being more popular, `requests` is a layer on top of `urllib3` itself, so it would be in any case a transitive dependency.

It is true that a stable and tested layer does a non-negligible job, but it seems like it is not allowing you to keep that `?` at the end. And if you need, there is no way around.
All in all, you are not doing anything too complicated with it, so it is perfectly fine as it is.